### PR TITLE
Elm 0.18

### PR DESCRIPTION
**Pull request by pzingg**
I gave myself an exercise to update your package to Elm 0.18.  I'm somewhat of an Elm newbie so I just could not get the "withFlags" and "withoutFlags" variants to compile following the way you extended AppWithFlags in your Elm 0.17 version. So you will see a lot of duplication of code.

The 5ef8746 commit actually compiled without all the unnecessary copying, but when I tried to run the examples in elm-reactor, I got this curious run-time message:

"Are you trying to sneak a Never value into Elm? Trickster!
It looks like Main.main is defined with `programWithFlags` but has type `Program Never`."

I guess because it doesn't the underlying program having flags, but being called as Program Never.

I also did a bit of copying in the examples, because the Elm 0.18 version of Http doesn't include the url method anymore. I grabbed the code from Elm 0.17 and put it inline in Examples 5, 6, and 7, renaming the method urlWithArgs.

### CODE REVIEW

**Comment by pzingg** [comment]
I refactored things in commit 6471019 to remove all the copying.  See if it works for you.

**Comment by pzingg** [comment]
I used elm-make --warn to generate the type annotations for the "Impl" functions that were not annotated in 6471019, so 3cd56fc now is fully annotated.

**Comment by fycth** [comment]
@rgrempel please, accept this PR!

**Comment by rgrempel** [comment]
This is impressive -- I had started on updating for Elm 0.18, but not gotten this far. Thanks!

I expect it will take me a few days to review all of this.

**Comment by pzingg** [comment]
I enjoyed using your package as a learning tool. i also dissected things and put them back another way--it's not as slick an encapsulation, but ends up exposing what is going on under the hood a little more. You can see that approach here: http://package.elm-lang.org/packages/pzingg/elm-navigation-extra/latest

It's probably more error-prone to someone forgetting to make certain "router" calls, but it doesn't add that much complexity to a stock `Navigation.program`.

File: README.md [hunk 1]
```diff
@@ -88,42 +88,28 @@ use elm-route-url, you don't have to.
 ### Mapping location changes to messages our app can respond to
 
 If you use the official [navigation](http://package.elm-lang.org/packages/elm-lang/navigation/latest)
-package directly, then you are asked to implement a function that acts
-sort of like `update`, but with a different signature -- it looks like this:
+package in Elm 0.18 directly, the `Navgation.program` differs from the
```
Comment (DavidHernandez) [inline README.md:91]: Small typo here. Should be `Navigation.program`.

**Comment by MarkNijhof** [comment]
How do I go about using the @pzingg repository until this (or something else) is merged?

**Comment by rgrempel** [comment]
Thanks again for your work on this!

I've pretty much got my additional changes ready to go on #22 -- I'll take one more look through that tomorrow and then hopefully publish the new version.

**Comment by mblarsen** [comment]
@pzingg @rgrempel great work.

Has it been published yet?

**Comment by rgrempel** [comment]
I got caught up in other things -- hopefully tomorrow.

**Comment by rgrempel** [comment]
Now published as version 3.0.

**Pull request merged**